charmcraft/models/project.py

```diff
@@ -129,7 +130,7 @@ def _validate_api_version(cls, value: str) -> str:
     @pydantic.field_validator("version", mode="before")
     def _validate_patch_version(cls, value: str) -> str:
         """Validate the optional patch version, providing a useful error message."""
-        api, separator, patch = value.partition(".")
+        api, separator, patch = str(value).partition(".")
```

Comment:
Strange that this conversion is necessary when it's declared as `str` in the function signature...

Reply:
It's because of the addition of `coerce_numbers_to_str` above. Sometimes pydantic will send the wrong type.